scripts/generate_matrix.py

```diff
@@ -0,0 +1,227 @@
+# !/usr/bin/env python3
+"""
+Script to generate a test matrix of Python and package versions for CI testing.
+
+This script fetches the list of supported Python versions and the versions of a
+specified package available on PyPI, determines compatible combinations based
+on package 'requires_python' metadata, and outputs a JSON matrix suitable for
+use in CI workflows.
+
+Features:
+- Generalized to work with any package (not just NumPy).
+- Caches HTTP requests to reduce network calls.
+- Prevents code duplication by using reusable functions.
+- Includes thorough doctests and documentation.
+
+Usage:
+    python generate_matrix.py [PACKAGE_NAME]
+
+Example:
+    python generate_matrix.py numpy
+
+"""
+
+import dataclasses
+import importlib.metadata
+import json
+import sys
+import urllib.error
+import urllib.request
+from functools import lru_cache
+
+# Constants for caching
+CACHE_SIZE = 128  # Adjust as needed
+
+
+@dataclasses.dataclass(frozen=True, eq=True, order=True)
```

Comment:
`eq=True` by default, and `slots` isn't

```suggestion
@dataclasses.dataclass(frozen=True, order=True, slots=True)
```

Reply:
Alternatively, you could also consider a `NamedTuple` here. That way, `minor_version` could be a simple slice. But it's also fine like this I suppose 🤷🏻

Reply:
Good plan!